### PR TITLE
posix: pthread: implement pthread_atfork()

### LINKED ISSUE #59934: posix: implement `pthread_atfork()`
There are currently 20 missing POSIX functions with trivial implementations, `pthread_atfork()` is one of them.

> pthread_atfork - register fork handlers
> 
> The pthread_atfork() function shall declare fork handlers to be called before and after [fork()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html), in the context of the thread that called [fork()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html). The prepare fork handler shall be called before [fork()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html) processing commences. The parent fork handle shall be called after [fork()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html) processing completes in the parent process. The child fork handler shall be called after [fork()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html) processing completes in the child process. If no handling is desired at one or more of these three points, the corresponding fork handler address(es) may be set to NULL.
> 
> If a [fork()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fork.html) call in a multi-threaded process leads to a child fork handler calling any function that is not async-signal-safe, the behavior is undefined.
> 
> The order of calls to pthread_atfork() is significant. The parent and child fork handlers shall be called in the order in which they were established by calls to pthread_atfork(). The prepare fork handlers shall be called in the opposite order.

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_atfork.html

See RFC #51211 for more info.

### PR DESCRIPTION
**Pull request by cfriedt**
pthread_atfork() is required by the POSIX_THREADS_BASE Option Group as detailed in Section E.1 of IEEE-1003.1-2017.

The POSIX_THREADS_BASE Option Group is required for PSE51, PSE52, PSE53, and PSE54 conformance, and is otherwise mandatory for any POSIX conforming system as per Section A.2.1.3 of IEEE-1003-1.2017.

Since Zephyr does not yet support processes and (by extension) fork(), this implementation includes a deviation and should be categorized as producing undefined behaviour.

Fixes #59934

### CODE REVIEW

**Review by kartben — DISMISSED** [review]

**Review by ycsin — DISMISSED** [review]

**Comment by cfriedt** [comment]
@kartben , @ycsin - sorry please re-ack. I just rebased (split some unrelated commits into #65642

**Comment by kartben** [comment]
> @kartben , @ycsin - sorry please re-ack. I just rebased (split some unrelated commits into #65642

Might be even better to just wait for posix docs change to be merged as this will conflict anyway, right?

**Review by kartben — APPROVED** [review]

**Review by ycsin — APPROVED** [review]

**Comment by cfriedt** [comment]
> Might be even better to just wait for posix docs change to be merged as this will conflict anyway, right?

A fast forward rebase with git managed to work fine. No merge conflict, and the .rst change made it to the right place in the end.

I was impressed

**Comment by kartben** [comment]
Please refrain from flagging such trivial-but-not-one-line either changes as trivial, especially if you're the one ending up merging. And trivial requires at least 4 hours in review before being merged anyway :)

**Comment by cfriedt** [comment]
> Please refrain from flagging such trivial-but-not-one-line either changes as trivial, especially if you're the one ending up merging. And trivial requires at least 4 hours in review before being merged anyway :)

That's ridiculous. This change was trivial.

If we're going to be so pedantic should I make  4 separate trivial PRs for each of the 4 lines of (directly related) changes in separate files?

**Pull request merged**
